### PR TITLE
[DOC] Add documentation entries for the execution policy interface of…

### LINKED ISSUE #2402: Document the intrerfaces for the parallel and vectorised alignments.
Currently, the execution policy interfaces for the alignment are not documented. 
We also need an tutorial for this.

### PR DESCRIPTION
**Pull request by rrahn**
fixes #2402

### CODE REVIEW

File: demos/dox/align/global_alignment_unbanded_execution_policy.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#include <iostream>
+
+#include <seqan/align_parallel.h>
+#include <seqan/stream.h>  // for printint strings
```
Comment (smehringer) [inline demos/dox/align/global_alignment_unbanded_execution_policy.cpp]:
```suggestion
#include <seqan/stream.h>  // for printing strings
```

File: demos/dox/align/global_alignment_unbanded_execution_policy.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    using TExecPolicy = seqan::ExecutionPolicy<TThreadModel, TVectorSpec>;
+
+    // dummy sequences
+    TSequence seqH = "CGATT";
+    TSequence seqV = "CGAAATT";
+
+    seqan::StringSet<TAlignedSequence> seqs1;
+    seqan::StringSet<TAlignedSequence> seqs2;
+
+    for (size_t i = 0; i < 100; ++i)
```
Comment (smehringer) [inline demos/dox/align/global_alignment_unbanded_execution_policy.cpp]:
```suggestion
    for (size_t i = 0; i < 100; ++i) // Create a data set of 100 dummy sequences
```

File: demos/dox/align/global_alignment_unbanded_execution_policy.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    setNumThreads(execPolicy, 4);
+
+    seqan::Score<int16_t, seqan::Simple> scoreAffine(2, -2, -1, -4);
+
+    seqan::String<int16_t> scores = seqan::globalAlignment(execPolicy, seqs1, seqs2, scoreAffine);
+
+    for (int16_t score : scores)
+        std::cout << "Score: " << score << "\n";
+
+    for (size_t pos = 0; pos < seqan::length(seqs1); ++pos)
```
Comment (smehringer) [inline demos/dox/align/global_alignment_unbanded_execution_policy.cpp]:
```suggestion
    for (size_t pos = 0; pos < seqan::length(seqs1); ++pos) // print out alignments
```

File: demos/dox/align/local_alignment_unbanded_execution_policy.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
+#include <iostream>
+
+#include <seqan/align_parallel.h>
+#include <seqan/stream.h>  // for printint strings
```
Comment (smehringer) [inline demos/dox/align/local_alignment_unbanded_execution_policy.cpp]:
```suggestion
#include <seqan/stream.h>  // for printing strings
```

File: demos/dox/align/local_alignment_unbanded_execution_policy.cpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    setNumThreads(execPolicy, 4);
+
+    seqan::Score<int16_t, seqan::Simple> scoreAffine(2, -2, -1, -4);
+
+    seqan::String<int16_t> scores = seqan::localAlignment(execPolicy, seqs1, seqs2, scoreAffine);
+
+    for (int16_t score : scores)
+        std::cout << "Score: " << score << "\n";
+
+    for (size_t pos = 0; pos < seqan::length(seqs1); ++pos)
```
Comment (smehringer) [inline demos/dox/align/local_alignment_unbanded_execution_policy.cpp]:
```suggestion
    for (size_t pos = 0; pos < seqan::length(seqs1); ++pos) \\ print out alignments
```

File: demos/dox/align/local_alignment_unbanded_score_execution_policy_parallel.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#include <iostream>
+
+#include <seqan/align_parallel.h>
+#include <seqan/stream.h>  // for printint strings
```
Comment (smehringer) [inline demos/dox/align/local_alignment_unbanded_score_execution_policy_parallel.cpp]:
```suggestion
#include <seqan/stream.h>  // for printing strings
```

File: demos/dox/align/local_alignment_unbanded_score_execution_policy_wavefront.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
+#include <iostream>
+
+#include <seqan/align_parallel.h>
+#include <seqan/stream.h>  // for printint strings
```
Comment (smehringer) [inline demos/dox/align/local_alignment_unbanded_score_execution_policy_wavefront.cpp]:
```suggestion
#include <seqan/stream.h>  // for printing strings
```

File: include/seqan/align/global_alignment_unbanded.h [hunk 8]
```diff
@@ -150,6 +161,39 @@ class Fragment;
...
  * https://seqan.readthedocs.io/en/develop/Tutorial/Algorithms/Alignment/PairwiseSequenceAlignment.html
  *
+ * @section Execution policies
+ *
+ * SeqAn supports parallel and vectorised execution of pairwise alignments.
+ * This additional interface takes not just a single pair of sequences but a collection of sequence pairs
+ * either in form of a collection over @link Align @endlink objects or as two collections of @link Gaps @endlink
+ * where both collections must have the same size. The collection based interface allows to additionally specify an
+ * @link ExecutionPolicy @endlink. This execution policy can be used to select the multi-threaded or vectorised
+ * implementation or the combination there of of the alignment algorithm. SeqAn implements an inter-sequence
```
Comment (smehringer) [inline include/seqan/align/global_alignment_unbanded.h]:
```suggestion
 * implementation or the combination of thereof for the alignment algorithm. SeqAn implements an inter-sequence
```

File: include/seqan/align/global_alignment_unbanded.h [hunk 9]
```diff
@@ -150,6 +161,39 @@ class Fragment;
...
+ * vectorisation scheme which means that <tt>x</tt> alignments are computed in parallel in one
+ * <a href="https://en.wikipedia.org/wiki/SIMD">SIMD</a> vector where <tt>x</tt> is the number of elements a vector
+ * can compute in parallel. This depends on the architecture's supported SIMD vector width (128 bit, 256 bit or 512 bit)
+ * and the selected score type, e.g. <tt>int16_t</tt>. For example on a CPU architecture that supports SSE4 and a score
+ * type of <tt>int16_t</tt>, <tt>128/16 = 8</tt> alignments can be computed in parallel on a single core.
+ * In addition, the execution policy can be configured for multi-threaded execution, such that chunks of sequence
+ * pairs from the initial collection are spawned and executed on different threads. The number of threads can be
+ * set via the execution policy.
+ *
+ * @note In order to get the performance advantage of vectorised execution one has to compile the code with the
```
Comment (smehringer) [inline include/seqan/align/global_alignment_unbanded.h]:
```suggestion
 * @note In order to get the performance advantage of vectorised execution, one has to compile the code with the
```

File: include/seqan/align/global_alignment_unbanded.h [hunk 10]
```diff
@@ -150,6 +161,39 @@ class Fragment;
...
+ *       respective CPU flags. For most Intel based CPUs the compiler flag <tt>-msse4</tt> can be used for gcc and
+ *       clang to enable vectorisation for 128 bit wide registers. For CPUs that support wider register please read
```
Comment (smehringer) [inline include/seqan/align/global_alignment_unbanded.h]:
```suggestion
 *       clang to enable vectorisation for 128 bit wide registers. For CPUs that support wider registers please read
```

**Comment by rrahn** [comment]
Many thanks for the quick apply suggestions!

File: demos/dox/align/global_alignment_unbanded_score_execution_policy_parallel.cpp [hunk 11]
```diff
@@ -0,0 +1,37 @@
+#include <iostream>
+
+#include <seqan/align_parallel.h>
+#include <seqan/stream.h>  // for printint strings
```
Comment (smehringer) [inline demos/dox/align/global_alignment_unbanded_score_execution_policy_parallel.cpp]:
```suggestion
#include <seqan/stream.h>  // for printing strings
```

File: demos/dox/align/global_alignment_unbanded_score_execution_policy_wavefront.cpp [hunk 12]
```diff
@@ -0,0 +1,46 @@
+#include <iostream>
+
+#include <seqan/align_parallel.h>
+#include <seqan/stream.h>  // for printint strings
```
Comment (smehringer) [inline demos/dox/align/global_alignment_unbanded_score_execution_policy_wavefront.cpp]:
```suggestion
#include <seqan/stream.h>  // for printing strings
```

File: include/seqan/align/local_alignment_unbanded.h [hunk 13]
```diff
@@ -141,6 +152,39 @@ namespace seqan {
...
  * @endcode
  *
+ * @section Execution policies
+ *
+ * SeqAn supports parallel and vectorised execution of local pairwise alignments.
+ * This additional interface takes not just a single pair of sequences but a collection of sequence pairs
+ * either in form of a collection over @link Align @endlink objects or as two collections of @link Gaps @endlink
+ * where both collections must have the same size. The collection based interface allows to additionally specify an
+ * @link ExecutionPolicy @endlink. This execution policy can be used to select the multi-threaded or vectorised
+ * implementation or the combination there of of the alignment algorithm. SeqAn implements an inter-sequence
```
Comment (smehringer) [inline include/seqan/align/local_alignment_unbanded.h]:
```suggestion
 * implementation or the combination thereof for the alignment algorithm. SeqAn implements an inter-sequence
```

**Comment by smehringer** [comment]
what bout the failing CI?

File: demos/dox/align/local_alignment_unbanded_execution_policy.cpp [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    setNumThreads(execPolicy, 4);
+
+    seqan::Score<int16_t, seqan::Simple> scoreAffine(2, -2, -1, -4);
+
+    seqan::String<int16_t> scores = seqan::localAlignment(execPolicy, seqs1, seqs2, scoreAffine);
+
+    for (int16_t score : scores)
+        std::cout << "Score: " << score << "\n";
+
+    for (size_t pos = 0; pos < seqan::length(seqs1); ++pos) \\ print out alignments
```
Comment (marehr) [inline demos/dox/align/local_alignment_unbanded_execution_policy.cpp]: this produces errors

File: demos/dox/align/local_alignment_unbanded_score_execution_policy_wavefront.cpp [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    TExecPolicy execPolicy;
+    seqan::setBlockSize(execPolicy, 500); // Sets the size of blocks used internally to partition the alignment matrix.
+    seqan::setParallelAlignments(execPolicy, 10); // Compute ten alignments at the same time.
+    seqan::setNumThreads(execPolicy, 4); // Use four threads to compute the actual alignment.
+
+    seqan::Score<int32_t, seqan::Simple> scoreAffine(2, -2, -1, -4);
+
+    seqan::String<int32_t> scores = seqan::localAlignmentScore(execPolicy, seqs1, seqs2, scoreAffine);
```
Comment (marehr) [inline demos/dox/align/local_alignment_unbanded_score_execution_policy_wavefront.cpp]: This seems to also fail.

include/seqan/align_parallel/async_wave_execution_interface.h:241:44: error: no matching function for call to ‘seqan::impl::AsyncWaveAlignExecutor<const seqan::String<seqan::SimpleType<unsigned char, seqan::Dna_>, seqan::Alloc<> >, const seqan::String<seqan::SimpleType<unsigned char, seqan::Dna_>, seqan::Alloc<> >, seqan::DPSettings<seqan::Score<int, seqan::Tag<seqan::Simple_> >, seqan::impl::doWaveAlignment(const seqan::ExecutionPolicy<seqan::WavefrontAlignment<TSpec>, TVectorizationSpec>&, const TAlgorithmSpec&, const TSetH&, const TSetV&, const TScore&, TArgs&& ...) [with TWaveSpec = seqan::Tag<seqan::BlockOffsetOptimization_>; TVectorizationPolicy = seqan::Tag<seqan::Vectorial_>; TAlgorithmSpec = seqan::LocalAlignment_<>; TSetH = seqan::StringSet<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna_>, seqan::Alloc<> > >; TSetV = seqan::StringSet<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna_>, seqan::Alloc<> > >; TScore = seqan::Score<int, seqan::Tag<seqan::Simple_> >; TArgs = {}; typename std::enable_if<((! std::is_same<seqan::WavefrontAlignment<TSpec>, seqan::Tag<seqan::Serial_> >::value) && (! std::is_same<seqan::WavefrontAlignment<TSpec>, seqan::Tag<seqan::Parallel_> >::value)), int>::type <anonymous> = 0]::DPConfigTraits> >::AsyncWaveAlignExecutor(const seqan::DPSettings<seqan::Score<int, seqan::Tag<seqan::Simple_> >, seqan::impl::doWaveAlignment(const seqan::ExecutionPolicy<seqan::WavefrontAlignment<TSpec>, TVectorizationSpec>&, const TAlgorithmSpec&, const TSetH&, const TSetV&, const TScore&, TArgs&& ...) [with TWaveSpec = seqan::Tag<seqan::BlockOffsetOptimization_>; TVectorizationPolicy = seqan::Tag<seqan::Vectorial_>; TAlgorithmSpec = seqan::LocalAlignment_<>; TSetH = seqan::StringSet<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna_>, seqan::Alloc<> > >; TSetV = seqan::StringSet<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna_>, seqan::Alloc<> > >; TScore = seqan::Score<int, seqan::Tag<seqan::Simple_> >; TArgs = {}; typename std::enable_if<((! std::is_same<seqan::WavefrontAlignment<TSpec>, seqan::Tag<seqan::Serial_> >::value) && (! std::is_same<seqan::WavefrontAlignment<TSpec>, seqan::Tag<seqan::Parallel_> >::value)), int>::type <anonymous> = 0]::DPConfigTraits>&, const seqan::ExecutionPolicy<seqan::WavefrontAlignment<seqan::Tag<seqan::BlockOffsetOptimization_> >, seqan::Tag<seqan::Vectorial_> >&)’

**Comment by rrahn** [comment]
> just some small stuff that broke and cdash said was broken.

Yes I know. I had no time to look into it yet. There is a lot in CI broken which is also not related. I will fix that too

**Pull request merged**